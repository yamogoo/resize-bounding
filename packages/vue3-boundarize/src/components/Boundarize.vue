<template>
  <div
    v-bind="$attrs"
    ref="refRoot"
    data-testid="boundarize-container"
    :class="[`${options.prefix}-${styles.container?.class}`]"
    :style="[computedStyle, styles?.container?.style]"
  >
    <slot :width :height></slot>
    <template v-for="({ show, direction }, idx) in panes" :key="idx">
      <BoundarizePane
        v-if="!disabled && show"
        :prefix="options.prefix ?? ''"
        :direction="direction"
        :options="options.pane"
        :styles="defaultStyles.pane"
        @focus="
          (isFocused) => {
            $emit(Emits.FOCUS, { state: isFocused, direction });
          }
        "
        @drag:start="onDragStart"
        @drag:move="onDragMove"
        @drag:end="onDragEnd"
      >
        <template #default>
          <slot name="knob"></slot>
        </template>
      </BoundarizePane>
    </template>
  </div>
</template>

<script lang="ts">
import {
  ref,
  computed,
  defineComponent,
  type PropType,
  type ComputedRef,
  type HTMLAttributes,
} from "vue";

import deepmerge from "deepmerge";

import type { BBResize } from "../shared/typings";

import BoundarizePane, {
  PaneDirectionAliases,
  PaneDirections,
  type PaneDirectionKey,
  type PaneEmittedData,
} from "./BoundarizePane.vue";

export enum Emits {
  FOCUS = "focus",
  UPDATE_WIDTH = "update:width",
  UPDATE_HEIGHT = "update:height",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export const defaultOptions: BBResize.Options = {
  prefix: "boundarize",
  pane: {
    width: 4,
    position: "central",
    knob: {
      show: false,
      constantlyShow: false,
    },
    showBorder: false,
    cursor: {
      horizontal: "col-resize",
      vertical: "row-resize",
    },
  },
};

export const defaultStyles: BBResize.Styles = {
  container: {
    class: "container",
    style: undefined,
  },
  pane: {
    class: "pane",
    style: {
      position: "absolute",
      display: "block",
      zIndex: 9999,
    },
    splitter: {
      class: "splitter",
      style: {
        position: "absolute",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        background: "none",
        transition: `background, height 125ms ease-out`,
        zIndex: 9999,
      },
      focused: {
        class: "active",
        style: { background: "blue" },
      },
    },
    knob: {
      class: "knob",
      style: {
        width: "64px",
        minWidth: "64px",
        height: "6px",
        background: "gray",
        borderRadius: "8px",
        transition: `background, height 125ms ease-out`,
      },
      focused: {
        class: "active",
        style: { background: "none" },
      },
    },
  },
};

export default defineComponent({
  setup(props, { emit }) {
    const refRoot = ref<HTMLDivElement | null>(null);

    let { width: newWidth, height: newHeight } = props;
    let prevWidth = newWidth,
      prevHeight = newHeight;

    const computedStyle: ComputedRef<HTMLAttributes["style"]> = computed(() => {
      const _width = {
        width: `${props.width}px`,
        minWidth: `${props.width}px`,
      };

      const _height = {
        height: `${props.height}px`,
        minHeight: `${props.height}px`,
      };

      return {
        position: "relative",
        ...(props.width && _width),
        ...(props.height && _height),
      };
    });

    const panes = computed(() => ({
      left: {
        show:
          RegExp(PaneDirections.LEFT).test(props.directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(props.directions),
        direction: PaneDirections.LEFT,
      },
      right: {
        show:
          RegExp(PaneDirections.RIGHT).test(props.directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(props.directions),
        direction: PaneDirections.RIGHT,
      },
      bottom: {
        show:
          RegExp(PaneDirections.BOTTOM).test(props.directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(props.directions),
        direction: PaneDirections.BOTTOM,
      },
      top: {
        show:
          RegExp(PaneDirections.TOP).test(props.directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(props.directions),
        direction: PaneDirections.TOP,
      },
    }));

    let startWidth = props.width ?? 0,
      startHeight = props.height ?? 0;

    let startX = 0,
      startY = 0;

    const options = computed(() => {
      return deepmerge(defaultOptions, props.options);
    });

    const styles: ComputedRef<BBResize.Styles> = computed(() => {
      return deepmerge(defaultStyles, props.styles) as BBResize.Styles;
    });

    const onDragStart = ({ x, y, dir }: PaneEmittedData): void => {
      startWidth = props.width ?? 0;
      startHeight = props.height ?? 0;

      startX = x;
      startY = y;

      emit(Emits.DRAG_START, dir);
    };

    const truncateInRange = (
      min: number,
      max: number | undefined,
      next: number,
    ): number => {
      const _max = max ?? Number.POSITIVE_INFINITY;

      if (next <= min) return min;
      else if (next >= _max) return _max;

      return next;
    };

    const onDragMove = ({ x, y, dir }: PaneEmittedData): void => {
      if (!refRoot.value) return;

      emit(Emits.DRAG_MOVE, dir);

      if (dir === PaneDirections.LEFT) {
        newWidth = startWidth + (startX - x);

        if (prevWidth === newWidth) return;

        const truncated = truncateInRange(
          props.minWidth,
          props.maxWidth,
          newWidth,
        );

        emit(Emits.UPDATE_WIDTH, truncated);
        prevWidth = truncated;
      } else if (dir === PaneDirections.RIGHT) {
        newWidth = startWidth + (x - startX);

        if (prevWidth === newWidth) return;

        const truncated = truncateInRange(
          props.minWidth,
          props.maxWidth,
          newWidth,
        );

        emit(Emits.UPDATE_WIDTH, truncated);
        prevWidth = truncated;
      } else if (dir === PaneDirections.TOP) {
        newHeight = startHeight + (startY - y);

        if (prevHeight === newHeight) return;

        const truncated = truncateInRange(
          props.minHeight,
          props.maxHeight,
          newHeight,
        );

        emit(Emits.UPDATE_HEIGHT, truncated);
        prevHeight = truncated;
      } else if (dir === PaneDirections.BOTTOM) {
        newHeight = startHeight + (y - startY);

        if (prevHeight === newHeight) return;

        const truncated = truncateInRange(
          props.minHeight,
          props.maxHeight,
          newHeight,
        );

        emit(Emits.UPDATE_HEIGHT, truncated);
        prevHeight = truncated;
      }
    };

    const onDragEnd = ({ dir }: PaneEmittedData): void => {
      emit(Emits.DRAG_END, dir);
    };

    return {
      refRoot,
      onDragStart,
      onDragMove,
      onDragEnd,
      truncateInRange,
      computedStyle,
      defaultStyles,
      options,
      styles,
      panes,
      PaneDirections,
      Emits,
    };
  },
  inheritAttrs: true,
  emits: {
    [Emits.FOCUS]: (data: { state: boolean; direction: string }) =>
      typeof data === "object" &&
      "state" in data &&
      typeof data.state === "boolean" &&
      "direction" in data &&
      typeof data.direction === "string",
    [Emits.UPDATE_WIDTH]: (width: number) => typeof width === "number",
    [Emits.UPDATE_HEIGHT]: (height: number) => typeof height === "number",
    [Emits.DRAG_START]: (dir: string) => typeof dir === "string",
    [Emits.DRAG_MOVE]: (dir: string) => typeof dir === "string",
    [Emits.DRAG_END]: (dir: string) => typeof dir === "string",
  },
  props: {
    width: Number,
    height: Number,
    minWidth: {
      type: Number,
      default: 0,
    },
    maxWidth: {
      type: Number,
      default: undefined,
    },
    minHeight: {
      type: Number,
      default: 0,
    },
    maxHeight: {
      type: Number,
      default: undefined,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    directions: {
      type: String as PropType<PaneDirectionKey | string>,
      default: `${PaneDirectionAliases.HORIZONTAL}${PaneDirectionAliases.VERTICAL}`,
    },
    alwaysShowKnob: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Object as PropType<Partial<BBResize.Options>>,
      default: {},
    },
    styles: {
      type: Object as PropType<Partial<BBResize.Styles>>,
      default: {},
    },
  },
  name: "Boundarize",
  components: { BoundarizePane },
});
</script>
