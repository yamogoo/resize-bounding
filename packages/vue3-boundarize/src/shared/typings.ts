import type { HTMLAttributes } from "vue";

export namespace BBResize {
  export interface Knob {
    show: boolean;
    constantlyShow: boolean;
  }

  export enum SplitterPositions {
    CENTER = "central",
    EXTERNAL = "external",
    INTERNAL = "internal",
  }

  export type PanePosition = SplitterPositions | string;

  export interface PaneOptions {
    width: number;
    position: PanePosition;
    showBorder: boolean;
    cursor: Partial<
      Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
    >;
    knob: Partial<Knob>;
  }

  export interface Options {
    prefix: string;
    pane: PaneOptions;
  }

  export interface StylingAttributes {
    class: string;
    style: HTMLAttributes["style"];
  }

  export type StylingStates = "focused";

  export type InteractiveElementStyles = Record<
    "splitter" | "knob",
    Record<StylingStates, StylingAttributes> & StylingAttributes
  >;

  export type PaneStyles = InteractiveElementStyles & StylingAttributes;

  export type Styles = Record<"container", StylingAttributes> & {
    pane: PaneStyles;
  };
}
