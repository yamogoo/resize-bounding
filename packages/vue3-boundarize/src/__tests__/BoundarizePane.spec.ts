import { describe, test, expect, beforeEach, vi } from "vitest";
import { shallowMount, VueWrapper } from "@vue/test-utils";
import { nextTick } from "vue";

import { DataTestIds, DEFAULT_PREFIX } from "./setup";

import deepmerge from "deepmerge";

import BoundarizePane, {
  Emits,
  PaneDirectionAliases,
  PaneDirections,
  paneBaseStyles,
  splitterBaseStyles,
  type PaneEmittedData,
  type Props,
} from "../components/BoundarizePane.vue";
import { BBResize } from "../shared/typings";

import { defaultOptions, defaultStyles } from "../components/Boundarize.vue";
// import { deepmerge } from "../utils";

const requiredProps = {
  styles: defaultStyles.pane,
  options: defaultOptions.pane,
};

describe("Boundarize", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("classes", () => {
    test.each(["my-prefix"])(
      "each element should have custom prefix (%s)",
      (prefix) => {
        const wrapper = shallowMount(BoundarizePane, {
          props: {
            prefix,
            ...requiredProps,
          },
        });

        const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
        const splitterEl = wrapper.find(
          `[data-testid="${DataTestIds.SPLITTER}"]`,
        );

        expect(paneEl.classes().toString()).toContain(prefix);
        expect(splitterEl.classes().toString()).toContain(prefix);

        expect(paneEl.classes()).toMatchSnapshot();
        expect(paneEl.classes()).toMatchSnapshot();
      },
    );

    test('should have "hide class by default"', () => {
      const wrapper = shallowMount(BoundarizePane, {
        props: {
          prefix: DEFAULT_PREFIX,
          ...requiredProps,
        },
      });

      const splitterEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

      expect(splitterEl.classes("hide")).toBeTruthy();
      expect(splitterEl.classes("hide")).toMatchSnapshot();
    });
  });

  describe("classes", () => {
    test('should have "show class"', async () => {
      const wrapper = shallowMount(BoundarizePane, {
        props: {
          prefix: DEFAULT_PREFIX,
          ...requiredProps,
        },
      });

      const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
      const splitterEl = wrapper.find(
        `[data-testid="${DataTestIds.SPLITTER}"]`,
      );

      await splitterEl.trigger("pointerenter");

      expect(paneEl.classes("show")).toBeTruthy();
      expect(paneEl.classes("show")).toMatchSnapshot();
    });
  });

  describe("events", () => {
    describe("focus", async () => {
      const wrapper = shallowMount(BoundarizePane, {
        props: {
          prefix: DEFAULT_PREFIX,
          direction: PaneDirections.RIGHT,
          ...requiredProps,
        },
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);

      test(`should emit "${Emits.FOCUS}" with state (true)`, async () => {
        await rootEl.trigger("pointerenter");
        const isFocused: boolean = (
          wrapper.emitted(Emits.FOCUS) as boolean[][]
        )[0][0];

        expect(isFocused).toBeTruthy();
        expect(isFocused).toMatchSnapshot();
      });

      test(`should emit "${Emits.FOCUS}" with state (false)`, async () => {
        await rootEl.trigger("pointerleave");
        const isFocused = (wrapper.emitted(Emits.FOCUS) as boolean[][])[1][0];

        expect(isFocused).toBeFalsy();
        expect(isFocused).toMatchSnapshot();
      });
    });

    describe("drag", () => {
      const wrapper = shallowMount(BoundarizePane, {
        props: {
          prefix: DEFAULT_PREFIX,
          direction: PaneDirections.RIGHT,
          ...requiredProps,
        },
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);

      test(`should emit "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" with {x, y, dir}`, async () => {
        const checkData = (
          eventName: string,
          wrapper: VueWrapper,
          eventData: { clientX: number; clientY: number },
        ): PaneEmittedData => {
          const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
          const data = ev[0][0];
          const { x, y } = data;

          const { clientX, clientY } = eventData;

          expect(data).toHaveProperty("x");
          expect(x).toBe(clientX);
          expect(data).toHaveProperty("y");
          expect(y).toBe(clientY);
          expect(data).toHaveProperty("dir");
          expect(data.dir).toBe(PaneDirections.RIGHT);
          return data;
        };

        let events: PaneEmittedData[] = [];

        const makeTrigger = async (
          event: string,
          eventData: { clientX: number; clientY: number },
          emittedEventName: string,
        ) => {
          await rootEl.trigger(event, eventData);
          events.push(checkData(emittedEventName, wrapper, eventData));
        };

        await rootEl.trigger("pointerenter", { clientX: 1, clientY: 1 });
        await makeTrigger(
          "pointerdown",
          { clientX: 1, clientY: 1 },
          Emits.DRAG_START,
        );
        await makeTrigger(
          "pointermove",
          { clientX: 2, clientY: 2 },
          Emits.DRAG_MOVE,
        );
        await makeTrigger(
          "pointerup",
          { clientX: 3, clientY: 3 },
          Emits.DRAG_END,
        );

        expect(events.length).toBe(3);
        expect(events).toMatchSnapshot();
      });
    });
  });

  describe("options", () => {
    describe("knob", () => {
      test("should always show knob", () => {
        const wrapper = shallowMount(BoundarizePane, {
          props: {
            prefix: DEFAULT_PREFIX,
            styles: defaultStyles.pane,
            options: deepmerge(defaultOptions.pane, {
              knob: {
                show: true,
                constantlyShow: true,
              },
            }),
          },
        });

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);

        expect(knobEl.exists()).toBeTruthy();
        expect(knobEl.attributes("data-testid")).toMatchSnapshot();
      });

      test.each([
        [PaneDirections.LEFT, 90],
        [PaneDirections.RIGHT, 90],
        [PaneDirections.TOP, 0],
        [PaneDirections.BOTTOM, 0],
      ])("pane (--%s) should rotated to (%s deg)", (direction, deg) => {
        const wrapper = shallowMount(BoundarizePane, {
          props: {
            prefix: DEFAULT_PREFIX,
            direction,
            styles: defaultStyles.pane,
            options: deepmerge(defaultOptions.pane, {
              knob: {
                show: true,
                constantlyShow: true,
              },
            }),
          },
        });

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);
        const styles = knobEl.attributes("style");

        expect(styles).toContain(`(${deg}deg)`);
        expect(styles).toMatchSnapshot();
      });
    });
  });

  describe("styles", () => {
    describe("cursor", () => {
      test('default cursor style should be "auto"', async () => {
        const wrapper = shallowMount(BoundarizePane, {
          props: {
            prefix: DEFAULT_PREFIX,
            ...requiredProps,
          },
        });
        await wrapper.setProps({
          direction: PaneDirections.RIGHT,
          prefix: DEFAULT_PREFIX,
        });

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);

        await rootEl.trigger("pointerenter");
        await rootEl.trigger("pointerleave");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain("auto");
        expect(cursorStyle).toMatchSnapshot();
      });

      const checkCursorActive = async (
        wrapper: VueWrapper,
        props: Partial<Props>,
        comparedCursorStyle: string,
      ): Promise<void> => {
        await wrapper.setProps(props);

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);
        await rootEl.trigger("pointerenter");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain(comparedCursorStyle);
        expect(cursorStyle).toMatchSnapshot();
      };

      describe("active", () => {
        describe("default styles", () => {
          test.each([
            PaneDirectionAliases.HORIZONTAL,
            PaneDirections.RIGHT,
            PaneDirections.LEFT,
          ])(
            'default cursor style should be "col-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(BoundarizePane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(wrapper, { direction }, "col-resize");
            },
          );

          test.each([
            PaneDirectionAliases.VERTICAL,
            PaneDirections.TOP,
            PaneDirections.BOTTOM,
          ])(
            'default cursor style should be "row-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(BoundarizePane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(wrapper, { direction }, "row-resize");
            },
          );
        });

        describe("custom styles (options)", () => {
          const CUSTOM_ACTIVE_CURSOR = "ew-resize";

          test.each([...Object.values(PaneDirections)])(
            `custom cursor style should be "${CUSTOM_ACTIVE_CURSOR}" (--%s)`,
            async (direction) => {
              const wrapper = shallowMount(BoundarizePane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(
                wrapper,
                {
                  direction,
                  options: {
                    cursor: {
                      vertical: CUSTOM_ACTIVE_CURSOR,
                      horizontal: CUSTOM_ACTIVE_CURSOR,
                    },
                    width: 0,
                    position: "",
                    showBorder: false,
                    knob: undefined,
                  },
                },
                CUSTOM_ACTIVE_CURSOR,
              );
            },
          );

          const SIZE = 12;
          const paneMap = paneBaseStyles(SIZE);
          const splitterMap = splitterBaseStyles(
            SIZE,
            BBResize.SplitterPositions.CENTER,
          );

          test.each([
            [PaneDirections.LEFT, paneMap.l],
            [PaneDirections.RIGHT, paneMap.r],
            [PaneDirections.TOP, paneMap.t],
            [PaneDirections.BOTTOM, paneMap.b],
          ])(
            `the pane (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(BoundarizePane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  direction,
                  styles: defaultStyles.pane,
                  options: { ...defaultOptions.pane, width: SIZE },
                },
              });

              await nextTick();

              const paneEl = wrapper.find(
                `[data-testid="${DataTestIds.PANE}"]`,
              );
              const style = paneEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );

          test.each([
            [PaneDirections.LEFT, splitterMap.l],
            [PaneDirections.RIGHT, splitterMap.r],
            [PaneDirections.TOP, splitterMap.t],
            [PaneDirections.BOTTOM, splitterMap.b],
          ])(
            `the pane splitter (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(BoundarizePane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  direction,
                  styles: defaultStyles.pane,
                  options: { ...defaultOptions.pane, width: SIZE },
                },
              });

              await nextTick();

              const splitterEl = wrapper.find(
                `[data-testid="${DataTestIds.SPLITTER}"]`,
              );
              const style = splitterEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );
        });
      });
    });

    describe("custom styles", () => {
      test('should apply custom styles for "pane", "splitter" and "knob" elements', async () => {
        const wrapper = shallowMount(BoundarizePane, {
          props: {
            prefix: DEFAULT_PREFIX,
            options: {
              ...defaultOptions.pane,
              knob: {
                show: true,
                constantlyShow: true,
              },
            },
            styles: deepmerge(defaultStyles.pane, {
              class: "pane",
              style: { border: "1px solid" },
              splitter: {
                class: "splitter",
                style: { background: "yellow" },
                focused: {},
                pressed: {},
              },
              knob: {
                class: "knob",
                style: { width: "320px", background: "pink" },
                focused: {},
                pressed: {},
              },
            }),
          },
        });

        const splitterEl = wrapper.find(
          `[data-testid="${DataTestIds.SPLITTER}"]`,
        );

        await splitterEl.trigger("pointerenter");
        const splitterStyles = splitterEl.attributes("style");

        const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
        const paneStyles = paneEl.attributes("style");

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);
        const knobStyles = knobEl.attributes("style");

        expect(paneStyles).toContain("border: 1px solid");
        // expect(splitterStyles).toContain("background: yellow"); // @!FIX
        expect(knobStyles).toContain("width: 320px");
        // expect(knobStyles).toContain("background: pink"); // @!FIX
      });

      test.each([
        [{ left: "70px", right: "70px", width: "70px" }, PaneDirections.RIGHT],
      ])(
        "inner styles of splitter should not have to overrided by customStyles",
        async (customStyles, direction) => {
          const SPLITTER_WIDTH = 12;

          const styles = splitterBaseStyles(
            SPLITTER_WIDTH,
            BBResize.SplitterPositions.CENTER,
          )[direction];

          const wrapper = shallowMount(BoundarizePane, {
            props: {
              prefix: DEFAULT_PREFIX,
              direction,
              styles: deepmerge(defaultStyles.pane, {
                splitter: {
                  style: customStyles,
                },
              }),
              options: { ...defaultOptions.pane, width: SPLITTER_WIDTH },
            },
          });

          const splitterEl = wrapper.find(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          await splitterEl.trigger("pointerenter");
          const splitterStyles = splitterEl.attributes("style");

          if (styles && splitterStyles) {
            for (const [k, v] of Object.entries(styles)) {
              const value = `${k}: ${v.toString()}`.trim();
              const trimedStyles = splitterStyles.toString().trim();

              expect(trimedStyles).toContain(value);
              expect(value).toMatchSnapshot();
              expect(trimedStyles).toMatchSnapshot();
            }
            expect(splitterStyles).toMatchSnapshot();
          }
        },
      );
    });
  });
});
