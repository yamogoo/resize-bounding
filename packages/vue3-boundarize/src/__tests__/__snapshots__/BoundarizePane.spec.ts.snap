// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Boundarize > classes > each element should have custom prefix (my-prefix) 1`] = `
[
  "my-prefix-pane",
  "--r",
  "hide",
]
`;

exports[`Boundarize > classes > each element should have custom prefix (my-prefix) 2`] = `
[
  "my-prefix-pane",
  "--r",
  "hide",
]
`;

exports[`Boundarize > classes > should have "hide class by default" 1`] = `true`;

exports[`Boundarize > classes > should have "show class" 1`] = `true`;

exports[`Boundarize > events > drag > should emit "drag:start -> drag:move -> drag:end" with {x, y, dir} 1`] = `
[
  {
    "dir": "r",
    "x": 1,
    "y": 1,
  },
  {
    "dir": "r",
    "x": 2,
    "y": 2,
  },
  {
    "dir": "r",
    "x": 3,
    "y": 3,
  },
]
`;

exports[`Boundarize > events > focus > should emit "focus" with state (false) 1`] = `false`;

exports[`Boundarize > events > focus > should emit "focus" with state (true) 1`] = `true`;

exports[`Boundarize > options > knob > pane (--b) should rotated to (0 deg) 1`] = `"width: 64px; min-width: 64px; height: 6px; background: gray; border-radius: 8px; transition: background, height 125ms ease-out; transform: rotate(0deg);"`;

exports[`Boundarize > options > knob > pane (--l) should rotated to (90 deg) 1`] = `"width: 64px; min-width: 64px; height: 6px; background: gray; border-radius: 8px; transition: background, height 125ms ease-out; transform: rotate(90deg);"`;

exports[`Boundarize > options > knob > pane (--r) should rotated to (90 deg) 1`] = `"width: 64px; min-width: 64px; height: 6px; background: gray; border-radius: 8px; transition: background, height 125ms ease-out; transform: rotate(90deg);"`;

exports[`Boundarize > options > knob > pane (--t) should rotated to (0 deg) 1`] = `"width: 64px; min-width: 64px; height: 6px; background: gray; border-radius: 8px; transition: background, height 125ms ease-out; transform: rotate(0deg);"`;

exports[`Boundarize > options > knob > should always show knob 1`] = `"boundarize-knob"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--b) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: ew-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--l) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: ew-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--r) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: ew-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--t) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: ew-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--b) should have "{ left: '0px', bottom: '6px', width: '100%', height: '0px' }" styles 1`] = `"position: absolute; display: block; z-index: 9999; left: 0px; bottom: 6px; width: 100%; height: 0px;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--l) should have "{ top: '0px', left: '6px', width: '0px', height: '100%' }" styles 1`] = `"position: absolute; display: block; z-index: 9999; top: 0px; left: 6px; width: 0px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--r) should have "{ top: '0px', right: '6px', width: '0px', height: '100%' }" styles 1`] = `"position: absolute; display: block; z-index: 9999; top: 0px; right: 6px; width: 0px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--t) should have "{ left: '0px', top: '6px', width: '100%', height: '0px' }" styles 1`] = `"position: absolute; display: block; z-index: 9999; left: 0px; top: 6px; width: 100%; height: 0px;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--b) should have "{ top: '0px', height: '12px', width: '100%' }" styles 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; background: none; top: 0px; height: 12px; width: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--l) should have "{ right: '0px', width: '12px', height: '100%' }" styles 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; background: none; right: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--r) should have "{ left: '0px', width: '12px', height: '100%' }" styles 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; background: none; left: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--t) should have "{ bottom: '0px', height: '12px', width: '100%' }" styles 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; background: none; bottom: 0px; height: 12px; width: 100%;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--h) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: col-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--l) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; right: 0px; width: 4px; height: 100%; cursor: col-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--r) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; width: 4px; height: 100%; cursor: col-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--b) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; top: 0px; height: 4px; width: 100%; cursor: row-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--t) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; bottom: 0px; height: 4px; width: 100%; cursor: row-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--v) 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; cursor: row-resize; background: blue;"`;

exports[`Boundarize > styles > cursor > default cursor style should be "auto" 1`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; width: 4px; height: 100%; cursor: auto; background: none;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 1`] = `"left: 0px"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 2`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; right: 70px; width: 12px; cursor: col-resize; background: blue; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 3`] = `"width: 12px"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 4`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; right: 70px; width: 12px; cursor: col-resize; background: blue; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 5`] = `"height: 100%"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 6`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; right: 70px; width: 12px; cursor: col-resize; background: blue; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 7`] = `"position: absolute; display: flex; align-items: center; justify-content: center; transition: background, height 125ms ease-out; z-index: 9999; left: 0px; right: 70px; width: 12px; cursor: col-resize; background: blue; height: 100%;"`;
